typeclasses/maps.py

```diff
@@ -0,0 +1,252 @@
+    The Map typeclass creates an empty map which fills up
+    as its owner travels through Ainneve.
+    When looking at the map, a representation of the world
+    as explored so far is shown.
+
+    Rooms can customize the map tile drawn for them in several ways:
+        - by specifying a `map_tile()` function that takes the requesting map
+          and returns a three-character string to draw on the map
+        - by specifying a `map_tile` property containing
+          a three-character string to draw on the map
+        - by specifying a `map_tile` Attribute containing
+          a three-character string to draw on the map
+        - by specifying a `terrain` present in the TILES dictionary in maps.py
+    If no tile can be found, the map will display `?!?` as the tile.
+    '''
+    weight = 1.0
+
+    def at_object_creation(self):
+        super(Map, self).at_object_creation()
+        self.db.map_data = {}
+        self.db.x = self.db.y = 0
+        self.db.off_map = False
+        self.map_current_room()
+
+    @property
+    def room(self):
+        '''
+        Finds the room that this object is currently in,
+        or None if it isn’t in a room.
+        This still works if the map is inside one or more nested objects.
+        '''
+        loc = self.location
+        if loc is None:
+            return loc
+
+        done = False
+        while 1:
+            if loc.location is not None:
+                loc = loc.location
+            else:
+                break
+        return loc
+
+    @property
+    def current_character(self):
+        '''
+        This returns the character carrying the map,
+        or None if the map isn’t currently being carried by a character.
+        '''
+        character = self.location
+        if character is None:
+            return character
+
+        done = False
+        while 1:
+            if character is None:
+                return None
+            if not character.is_typeclass('characters.Character'):
+                character = character.location
+            else:
+                break
+        return character
+
+    def draw_map(self, show_player = True):
+        '''
+        This returns a two-dimensional list containing the map tiles.
+
+        If the owner is a builder, the map displays rows
+        containing the DB numbers for each room above each row.
+
+        If the show_player argument is `True`, the room the owning player
+        is currently in will have its middle character replaced with a `*`.
+        '''
+        player = self.current_character.player
+        is_builder = player.locks.check_lockstring(player, 'dummy:perm(Builders)')
+
+        min_x = max_x = 0
+        min_y = max_y = 0
+        for room_info in self.db.map_data.values():
+            x = room_info['x']
+            y = room_info['y']
+            if x < min_x: min_x = x
+            if y < min_y: min_y = y
+            if x > max_x: max_x = x
+            if y > max_y: max_y = y
```

Comment:
Lines 93 and 94 should possibly be split to two lines each.
Each of the if statements should also possibly be split:
```
if x < min_x:
    min_x = x
```
Again, for clarity and easy reading.